### DISCUSSION TITLE
major performance degradations from using PRNG on TPU

### DISCUSSION THREAD

**Discussion by dlwh** [jax-ml/jax#12001] [Q&A]
Hi,

I'm continuing to benchmark/understand a baseline transformer implementation on TPU. I've found that dropout causes a very significant performance penalty (as much as a factor of 3 slowdown). I then tried switching to the hardware RNG, following the recipe I've seen in https://github.com/google/jax/issues/9274 and in t5x and this results in a further order of magnitude penalty.

I'm guessing at least the hardware RNG slowdown isn't expected? Has anyone run into this?

**Comment by froystig** [comment]
I'm not sure that I'd expect it with the XLA "RandomBitGenerator" RNG. Looking at the recipe in #9274 you linked, the suggestion involves patching up modules, which isn't super robust. There's a better way to spell it anyway; just change the process-wide RNG implementation:
```python
jax.config.update('jax_default_prng_impl', 'unsafe_rbg')
```
Then no need to patch anything up nor use `jax.prng.seed_with_impl`, just continue using `jax.random.PRNGKey(seed)` and the rest of `jax.random`.

And if you need more dynamic control of the default RNG implementation, you can also use:
```python
with jax.default_prng_impl('unsafe_rbg'): ...
```
but beware mixing keys coming from within/without such a dynamic scope.

Flipping on `config.jax_enable_custom_prng` is still experimental and there are known bugs when you do that, so I wouldn't recommend it quite yet. We will flip its default value and deprecate the `False` state when we know that it's ready.

How do things look when you switch RNGs this way? Any better?

**Reply by dlwh** [reply]
Sorry for being slow and thanks for the response! I ended up giving up on it for the time being, but I'll get to it soon!

**Reply by dlwh** [reply]
Ok I took a look this morning. tl; dr: it didn't help, or didn't help much...

This is a more or less faithful reimplementation of gpt2-small on a v3-32, one example per tpu (which isn't the most I can fit but...), measuring time to process a single batch of 512 examples (16 grad accum steps):

* No dropout: 1s
* dropout, default prng: 2.5s
* dropout, unsafe_rbg: 5.2s

This is based on jax main as of last night.

Happy to share code, look at dumps from XLA, etc!

Thanks!

**Comment by jekbradbury** [comment]
If you're using pjit on TPU, make sure to also enable `LIBTPU_INIT_ARGS=--xla_tpu_spmd_rng_bit_generator_unsafe=1`. Without that, XLA SPMD can't split the RBG op up because it would change numerics.

**Reply by dlwh** [reply]
Thanks for the suggestion, but it made no difference at all. (Just to be sure, I verified that I got the env var set by making sure libtpu got mad when i gave it a garbage flag)

**Reply by jekbradbury** [reply]
Hmm :(
Maybe the issue is lack of fusion, then? Any chance you can take a profile and upload the `trace.json.gz` file?

**Reply by dlwh** [reply]
Gladly! But I think I messed something up. I put this in my training loop:

```
          if step == 5:
                jax.profiler.start_trace("/tmp/gpt2-trace")
            elif step == 105:
                jax.profiler.stop_trace()
```

and it's definitely past step 105, and all I see in that dir is a 40 byte file named `events.out.tfevents.1661366413.t1v-n-6ac7fb9e-w-2.profile-empty`. Same on all hosts.

**Reply by dlwh** [reply]
I should mention I got this warning when I started it:

```
W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: :/usr/local/lib
```

(I'm definitely not on a CUDA machine!)

**Reply by dlwh** [reply]
not sure what I did wrong the first time, but here it is!
[t1v-n-6ac7fb9e-w-3.trace.json.gz](https://github.com/google/jax/files/9418760/t1v-n-6ac7fb9e-w-3.trace.json.gz)

**Reply by jekbradbury** [reply]
Interesting; I see a lot of tiny op-by-op computations here, as would happen if you're not jitting/pjitting your model at all. (It's possible that the trace simply didn't capture a step of the compiled model, but all I can see is op-by-op code.)

**Reply by dlwh** [reply]
I seem to recall not seeing anything after the first step when I first tried profiling a while back (and haven't messed with it since). That might explain why when I tried to run a profile from step 5 to 105 it didn't capture anything at all... Maybe it just isn't capturing anything once the jit has run?

I'm happy to share the xla dump if that's helpful. I stared at it some and didn't see anything that stood out (well, related to dropout), but I also don't know I'm looking for.

**Reply by dlwh** [reply]
I looked at the xla output for the hardware. This is what I see:

This is the block that generates the needed random bits, and I think maybe it explains what's going on... I'm getting a bit lost, but it looks like it is indeed generating the masks in a loop, one for each node.

The `u32[1,12,1024,1024]` is the precursor to the floats that get generated for the dropout mask. The "while" runs until an accumulator hits 32 (i.e. the number of tpus), though I'm not sure what the increment and start are, but I'm guessing they're 0 and 1 (I just got very lost following all the dataflow.)

I can't find an analogue to this in the non-hardware prng code. It looks fairly direct.

The env variables I used are:

`LIBTPU_INIT_ARGS=--xla_tpu_spmd_rng_bit_generator_unsafe=1 XLA_FLAGS="--xla_dump_to=/tmp/xla_dumps/hardware_rng/ --xla_dump_hlo_pass_re=.*`

```
 while.73 = (s32[]{:T(256)}, u32[1,12,1024,1024]{3,2,1,0:T(8,128)}, u32[32,4]{1,0:T(8,128)}, s32[]{:T(256)}, s32[]{:T(256)}, /*index=5*/s32[]{:T(256)}) while(tuple.623), condition=wide.region_23.805_spmd.clone, body=wide.region_22.774_spmd.clone

wide.region_23.805_spmd.clone {
  wide_param.6 = (s32[]{:T(256)}, u32[1,12,1024,1024]{3,2,1,0:T(8,128)}, u32[32,4]{1,0:T(8,128)}, s32[]{:T(256)}, s32[]{:T(256)}, /*index=5*/s32[]{:T(256)}) parameter(0)
  get-tuple-element.4147 = s32[]{:T(256)} get-tuple-element(wide_param.6), index=0
  constant.3375 = s32[]{:T(256)} constant(32)
  ROOT compare.269 = pred[]{:T(1024)} compare(get-tuple-element.4147, constant.3375), direction=LT, metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/cond/lt" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
}

wide.region_22.774_spmd.clone {
  wide_param.7 = (s32[]{:T(256)}, u32[1,12,1024,1024]{3,2,1,0:T(8,128)}, u32[32,4]{1,0:T(8,128)}, s32[]{:T(256)}, s32[]{:T(256)}, /*index=5*/s32[]{:T(256)}) parameter(0)
  get-tuple-element.5619 = s32[]{:T(256)} get-tuple-element(wide_param.7), index=0
  copy.179 = s32[]{:T(256)S(6)} copy(get-tuple-element.5619)
  get-tuple-element.5628 = s32[]{:T(256)} get-tuple-element(wide_param.7), index=3
  add.1093 = s32[]{:T(256)} add(copy.179, get-tuple-element.5628), metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/add" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  get-tuple-element.5620 = u32[1,12,1024,1024]{3,2,1,0:T(8,128)} get-tuple-element(wide_param.7), index=1
  get-tuple-element.5627 = u32[32,4]{1,0:T(8,128)} get-tuple-element(wide_param.7), index=2
  get-tuple-element.5629 = s32[]{:T(256)} get-tuple-element(wide_param.7), index=4
  compare.270 = pred[]{:T(1024)S(6)} compare(copy.179, get-tuple-element.5629), direction=LT, metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/lt" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  get-tuple-element.5630 = s32[]{:T(256)} get-tuple-element(wide_param.7), index=5
  add.1094 = s32[]{:T(256)S(6)} add(copy.179, get-tuple-element.5630), metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/add" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  select.461 = s32[]{:T(256)S(6)} select(compare.270, add.1094, copy.179), metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/select_n" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  fusion.690 = u32[4]{0:T(256)} fusion(get-tuple-element.5627, select.461), kind=kLoop, calls=fused_computation.122.clone
  reshape.1479 = u32[2,2]{1,0:T(2,128)} reshape(fusion.690), metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/rng_bit_generator[shape=(12, 1024, 1024) dtype=uint32 algorithm=RNG_DEFAULT]" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  fusion.691 = (u32[2]{0:T(256)}, u32[2]{0:T(256)}) fusion(reshape.1479), kind=kLoop, calls=fused_computation.120.clone, metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/rng_bit_generator[shape=(12, 1024, 1024) dtype=uint32 algorithm=RNG_DEFAULT]" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  rng-bit-generator.24 = u32[12,1024,1024]{2,1,0:T(8,128)} rng-bit-generator(fusion.691), algorithm=rng_default, metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/rng_bit_generator[shape=(12, 1024, 1024) dtype=uint32 algorithm=RNG_DEFAULT]" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  partition-id.40 = u32[]{:T(256)} partition-id(), metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/dynamic_update_slice" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  convert.811 = s32[]{:T(256)S(6)} convert(partition-id.40), metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/dynamic_update_slice" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  compare.273 = pred[]{:T(1024)S(6)} compare(select.461, convert.811), direction=GE, metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/dynamic_update_slice" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  add.1098 = s32[]{:T(256)S(6)} add(convert.811, get-tuple-element.5628), metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/dynamic_update_slice" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  compare.274 = pred[]{:T(1024)S(6)} compare(select.461, add.1098), direction=LT, metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/dynamic_update_slice" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  and.452 = pred[]{:T(1024)S(6)} and(compare.273, compare.274), metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/dynamic_update_slice" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  fusion.692 = u32[1,12,1024,1024]{3,2,1,0:T(8,128)} fusion(get-tuple-element.5620, rng-bit-generator.24, and.452), kind=kLoop, calls=fused_computation.106.clone, metadata={op_name="pjit(train_step)/jit(main)/while/body/jvp(vmap(Gpt2Transformer))/while/body/Gpt2Block/Gpt2Attention/dropout/vmap(while)/body/dynamic_update_slice" source_file="/files/levanter/src/levanter/nn/static_dropout.py" source_line=62}
  ROOT tuple.625 = (s32[]{:T(256)}, u32[1,12,1024,1024]{3,2,1,0:T(8,128)}, u32[32,4]{1,0:T(8,128)}, s32[]{:T(256)}, s32[]{:T(256)}, /*index=5*/s32[]{:T(256)}) tuple(add.1093, fusion.692, get-tuple-element.5627, get-tuple-element.5628, get-tuple-element.5629, /*index=5*/get-tuple-element.5630)
}```

**Reply by dlwh** [reply]
fwiw, i decided to revisit this today, and I now get a segfault when using the flag`--xla_tpu_spmd_rng_bit_generator_unsafe=1`

I tried:
jax==0.3.17
libtpu-nightly==0.1.dev20220723

jax==0.3.17
libtpu-nightly-0.1.dev20220913

jax main as of today: git+https://github.com/google/jax/
libtpu-nightly-0.1.dev20220913

(Though possibly this is the thing that's going to land soon-ish @froystig ?)

**Reply by froystig** [reply]
What I hope to put together soon-ish is a more partitioning-friendly variant of the threefry RNG (rather than anything with the XLA bit generator). Segfaults are no good! Hopefully we can debug that regardless...

**Reply by dlwh** [reply]
ok I've minimized it. Looks like it's a libtpu arg parsing bug since it crashes before/during device initialization. I'll file a bug with xla?

**Reply by dlwh** [reply]
oh user error (but also I guess segfaults are always bad!) I was passing in via XLA_FLAGS and not LIBTPU_INIT_ARGS ...

**Reply by froystig** [reply]
Ah, good that you've figured this out. Still, not worth a segfault if you ask me! I'll let XLA know.

**Reply by dlwh** [reply]
i sent a bug to the trc-support front door.
